### PR TITLE
Fix broken tests

### PR DESCRIPTION
**Pull request by gjcairo**
### Motivation:

This PR fixes several broken/flaky unit and integration tests, that started failing in our nightly pipeline. 

Namely:
- `NIOThrowingAsyncSequenceTests.testIteratorThrows_whenCancelled`: unit test was flaky.
- `EventLoopCrashTests.testInstallingSingletonMTELGAsConcurrencyExecutorWorksButOnlyOnce`: integration test failed in nightly.
- Compilation of allocation tests in nightly was failing.

### Modifications:

- Modified flaky unit test by fixing a small, edge-case bug (see [this comment](https://github.com/apple/swift-nio/pull/2621#discussion_r1443106360) for details).
- `EventLoopCrashTests.testInstallingSingletonMTELGAsConcurrencyExecutorWorksButOnlyOnce` was failing because we forgot to disable the test for Swift versions greater than 5.10. There isn't sufficient testing done on this feature on versions greater than 5.10, so the whole feature is simply disabled in those cases, so the test should be too.
- Building allocation tests was failing in nightly because `AtomicCounter` was not an explicit dependency of the test targets. This worked before because all modules were always visible to targets, but this changed in https://github.com/apple/swift-package-manager/pull/7103. I added `AtomicCounter` as a dependency to these targets and the build now succeeds.
- Updated old `swift-tools-version` in generated `Package.swift`s to 5.7. This wasn't necessary to fixing the issues but they were lagging behind by a lot (some were version 5.0).

### Result:

Happy nightly pipeline.

### CODE REVIEW

File: Tests/NIOCoreTests/AsyncSequences/NIOThrowingAsyncSequenceTests.swift [hunk 1]
```diff
@@ -871,7 +872,19 @@ final class NIOThrowingAsyncSequenceProducerTests: XCTestCase {
...
+                            // This loop will exit when the task is cancelled
+                            // or when all of the elements have been iterated.
+                            // It is possible that the cancellation will be
+                            // triggered after the last element was yielded,
+                            // just before the iterator can return the `nil`
+                            // signifying the end of the sequence.
+                            // If this happens, we avoid incrementing the counter,
+                            // so that we don't increase it to a value beyond the
+                            // maximum possible sequence value.
+                            counter += 1
```
[the remark this answers is not in the document]
Comment (gjcairo) [reply Tests/NIOCoreTests/AsyncSequences/NIOThrowingAsyncSequenceTests.swift:886]: We're not counting above the size of the array - the problem is with how we're indexing. By the time the sequence yields 99, `counter` will also have a value of 99. However, at the end of the iteration, we increment the counter to 100, bringing us past the limit and failing the assertion `XCTAssertLessThan(counter, 100)`.

The confusion comes from the fact that `counter` is counting elements yielded/iterations completed, but we assert its value matches the yielded value from the sequence, which starts at 0.

There are several ways of fixing this, and I agree this was probably not very clear. I've changed it to something that is hopefully more obvious now.

**Review by Lukasa — APPROVED** [review]
